### PR TITLE
HTML-381: Added validation while processing obs and obsgroup

### PR DESCRIPTION
**Pull request by geoff-wasilwa**
HFE did not provide checks for obs used as set members and obsgroup (if the obsgroup concept was a set).
Added a mechanism to validate this and store concept ids violating this rule to FormEntryContext from where
they can be displayed on the HTML Form edit page.

### CODE REVIEW

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsGroupTagHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    public void before() throws Exception {
+    	
+    	executeDataSet(XML_OPENMRS_TEST_DATASET);
+    	
+    	Patient demo = HtmlFormEntryUtil.getFakePerson();
+        formEntrySession = new FormEntrySession(demo, XML_DEFINITION, null);
+
+        out = new CapturingPrintWriter();
+        handler = new ObsGroupTagHandler();
+
```
Comment (mogoodrich) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsGroupTagHandlerTest.java]: Can you clean up the tabs/formatting in this file? Thanks!

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsGroupTagHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    public void testSettingNonSetConceptAsObsGroupConcept() throws Exception {
+    	
+    	Document document = HtmlFormEntryUtil.stringToDocument(XML_DEFINITION);
+        Node obsgroupNode = HtmlFormEntryUtil.findDescendant(document, "obsgroup");
+        assertTrue(handler.doStartTag(formEntrySession, out, obsgroupNode.getParentNode(), obsgroupNode));
+        assertEquals(1, formEntrySession.getContext().getInvalidObsGroupConceptIds().size());
+    	
+    }
+	
+}
```
Comment (mogoodrich) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsGroupTagHandlerTest.java]: Can you add a test to make sure that we *don't* get an error if the obs group is a set?

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsTagHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	
+	@Test
+	public void testAddingNonSetMembersToObsGroup() throws Exception {
+		
+		Document document = HtmlFormEntryUtil.stringToDocument(XML_DEFINITION);
+		Node obsGroupNode = HtmlFormEntryUtil.findDescendant(document, "obsgroup");
+        Node obsNode = HtmlFormEntryUtil.findDescendant(document, "obs");
+        Boolean digDeeper = obsGroupHandler.doStartTag(formEntrySession, out, obsGroupNode.getParentNode(), obsGroupNode);
+        Boolean digDeeper2 = obsHandler.doStartTag(formEntrySession, out, obsNode.getParentNode(), obsNode);
+        assertEquals(1, formEntrySession.getContext().getInvalidSetMemberConceptIds().size());
```
Comment (mogoodrich) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsTagHandlerTest.java]: Nitpick, but if we aren't using the "dipDeeper" variables, can we just remove them?  The meaning of "digDeeper" isn't obviously apparent.

Comment (mogoodrich) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsTagHandlerTest.java]: Also, like in the above test, can we add an confirmation that we getInvalidSetMemberConceptIds() is null/0 if the form is properly defined?

File: api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java [hunk 4]
```diff
@@ -96,6 +98,10 @@
     // TODO 1) change the type of this variable to visit
     // TODO 2) change HtmlFormEntryController so that it correctly populates the context with the relevent visit (if available)
     private Object visit;
+    
+    private List<Integer> invalidObsGroupConceptIds = new ArrayList<Integer>();
+    
+    private Map<Integer,List<Integer>> invalidSetMemberConceptIds = new HashMap<Integer,List<Integer>>();
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java]: It seems kind of odd to make this error counters a property of the Form Entry Context, though the "right" location doesn't spring to mind--did you think about other places to handle this?

File: api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java [hunk 5]
```diff
@@ -44,6 +47,18 @@ public boolean doStartTag(FormEntrySession session, PrintWriter out, Node parent
...
+		ObsGroup currentObsGroup = context.getActiveObsGroup();
+		Map<String, String> attributes = getAttributes(node);
+		if (currentObsGroup != null && currentObsGroup.getConcept().isSet() && attributes.containsKey("conceptId")) {
+			Concept obsConcept = HtmlFormEntryUtil.getConcept(attributes.get("conceptId"));
+			if (!currentObsGroup.getConcept().getSetMembers().contains(obsConcept)) {
+				context.addInvalidSetMemberConceptIds(currentObsGroup.getConcept().getConceptId(), obsConcept.getConceptId());
+			}
+		}
+	}
+
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java]: I'd have preferred that this (and the validation test in ObsGroupTagHandler) happened in HtmlFormValidator, instead of in the tag handlers themselves, because we really only need to perform this logic a small fraction of the time (when an admin is editing a form schema, not when end users are entering/editing a form). Did you consider this at all?  I do realize that it would require some significant changes/improvements to HtmlFormValidator.

File: omod/src/main/java/org/openmrs/module/htmlformentry/web/controller/HtmlFormController.java [hunk 6]
```diff
@@ -80,9 +81,16 @@ public void showHtmlForm(Model model,
...
                 if (fes.getFieldAccessorJavascript() != null) {
                 	html += "<script>" + fes.getFieldAccessorJavascript() + "</script>";
                 }
+				if (!formEntryContext.getInvalidObsGroupConceptIds().isEmpty()) {
+					model.addAttribute("invalidObsGroupConceptIds", formEntryContext.getInvalidObsGroupConceptIds());
+				}
+				if (!formEntryContext.getInvalidSetMemberConceptIds().isEmpty()) {
+					model.addAttribute("invalidSetMemberConceptIds", formEntryContext.getInvalidSetMemberConceptIds());
+				}
                 model.addAttribute("previewHtml", html);
```
Comment (mogoodrich) [inline omod/src/main/java/org/openmrs/module/htmlformentry/web/controller/HtmlFormController.java]: Shouldn't we also be testing when calling the saveHtmlForm method?

File: api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java (hunk 5 above)
Comment (djazayeri) [inline api/src/main/java/org/openmrs/module/htmlformentry/handler/ObsTagHandler.java]: +1 to this comment. This should really be tested at time of saving the form definition. It is not useful to test this when a regular user is filling it out, and showing them a warning is only going to be confusing, because what can they do at that point?

**Comment by geoff-wasilwa** [comment]
:) thanks, your comments really got me thinking. Working on moving the validations to HtmlFormValidator.

File: api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java (hunk 4 above)
Comment (geoff-wasilwa) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java]: What do you think of storing the error counter in HtmlFormValidator?

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsTagHandlerTest.java (hunk 3 above)
Comment (geoff-wasilwa) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsTagHandlerTest.java]: 'digDeeper' is a side effect of copy paste :)

File: api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java (hunk 4 above)
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java]: Having the HtmlFormValidator handle the validation makes sense... if using an error counter within it is the best way to go, then yes.  Not sure if moving to the validator will lead you to change your design complete though.

File: api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsTagHandlerTest.java (hunk 3 above)
Comment (mogoodrich) [inline api-tests/src/test/java/org/openmrs/module/htmlformentry/handler/ObsTagHandlerTest.java]: Yeah, after posting that comment, I saw the digDeeper in the original code.

File: api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java (hunk 4 above)
Comment (djazayeri) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java]: If the warnings will be generated by HtmlGeneratorElement#generateHtml(FormEntryContext) then it makes sense to track warnings here in FormEntryContext, because this is really the "HTML Generation Context". I don't like hardcoding a holder for just this one specific validation. I would expect the property to be

    List<FormGenerationWarning>

and there can be

    class ObsGroupWithInvalidMember implementes FormGenerationWarning

However that means that the warnings will be generated every time the form is used, which could be an unnecessary performance hit.

My own inclination would be to do something like introducing a new sub-interface that gives more details about compiling (because we can't directly change HtmlGenerationElement for backwards-compatibility)

    interface AnalyzableHtmlGeneratorElement extends HtmlGeneratorElement {
        HtmlGenerationAnalysis analyze(FormEntryContext)
    }

Comment (geoff-wasilwa) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java]: Thanks @djazayeri, was struggling with the implementation of the warnings. Although I have changed the design, such that HtmlFormValidator creates a Document object from HtmlForm's xmlData and traverses the document validating obsgroup and obs element. 

I also had to change the stringToDocument method of HtmlFormEntryUtil so as to add line and column numbers as per this [blog](https://eyalsch.wordpress.com/2010/11/30/xml-dom-2/) and this [SO question](http://stackoverflow.com/questions/4915422/get-line-number-from-xml-node-java).

What's your take on this?

Comment (geoff-wasilwa) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java]: Aside: the initial plan was to have the warnings as a string collection in HtmlFormValidator

Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/FormEntryContext.java]: Offhand, having the validator create a Document object from the xmlData makes sense to me--opens up the possibility of adding other document-based validations in the future.

I just glanced at the blog and question you linked to, but i assume you want to include line numbers so that you can point the end user to where the actual error is in the file?  That does sound helpful.  How are you planning on doing this?

File: api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java [hunk 7]
```diff
@@ -66,4 +90,117 @@ private boolean hasEncounterTypeTag(String xml) throws Exception {
 		Node formNode = HtmlFormEntryUtil.findChild(doc, "htmlform");
 		return HtmlFormEntryUtil.findChild(formNode, "encounterType") != null;
 	}
+    
+	private void setHtmlFormWarnings(String warning) {
+		htmlFormWarnings.add(warning);
+	}
+	
+	private void traverseNode(Node currentNode, Node parentObsGroupNode, Errors errors) {
```
Comment (geoff-wasilwa) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: @mogoodrich this is what I had in mind, although I've been having internal arguments regarding implementation of the validators (it might grow out of control)

Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: @geoff-wasilwa looks like a good start to me. I agree that this could grow into a long if/then block as the number of validators grows, which would be ugly.  I'm not *too* concerned about it if you want to do it this way, but, yeah, if you could figure out a better lookup method for validators, that would be cool.  Potentially, even, each TagHandler could implement the validation for that tag within the class?  Then you could model something after the applyTagsHandler methods in HtmlFormEntryGenerator. Actually, maybe you could add a new interface TagValidator, that a TagHandler could optionally implement? (It would have to be optional at this point because there are TagHandlers defined outside of the HFE module that we wouldn't want to break).

Does that make sense? @djazayeri, let me know if you think I'm going down the wrong path.

Comment (djazayeri) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: Yes, I agree with a new subinterface of TagValidator (and that we have to
make this optional because of backwards-compatibility).

I was going to comment along these lines earlier. Maybe I did? Or maybe I
never posted that comment...

Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: Darius, looks like you did make the comment.  Do you think this sub-interface should be on the TagHandler or the HtmlGeneratorElement?

Comment (djazayeri) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: Offhand I'm not sure. I didn't look closely when saying HtmlGeneratorElement.

Comment (geoff-wasilwa) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: Okay guys, thanks for the comments let me work on the changes.

File: api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java [hunk 8]
```diff
@@ -16,6 +25,8 @@
 
 	protected final Log log = LogFactory.getLog(getClass());
 	
+	private List<String> htmlFormWarnings = new ArrayList<String>();
```
Comment (geoff-wasilwa) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: Will the string implementation of the warnings work or should I go with the earlier suggestion of having a `List` of `FormGenerationWarning`?

Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormValidator.java]: Seems like it would be nice to have some sort of object that in addition to the string also includes the element tag, line number, etc... but I'd be fine with just string as well...

**Comment by geoff-wasilwa** [comment]
Sorry I took long to deliver on this. Got a few issues though:
- can I use the .warn class found in uicommons module to represent the warnings?
- will I have to bind the warnings to request (`request.setAttribute(..)`) so that the warnings are displayed once the form has been saved?

**Comment by geoff-wasilwa** [comment]
@mogoodrich and @djazayeri any thoughts?

**Comment by geoff-wasilwa** [comment]
@mogoodrich I have made the changes. Thanks for pointing them out.

**Comment by mogoodrich** [comment]
Great!  I will merge this is in now and then try to test it out on some of our forms.  Thanks!
